zerver/actions/message_send.py

```diff
@@ -488,6 +490,9 @@ def get_ids_for(f: Callable[[ActiveUserDict], bool]) -> set[int]:
     # where we determine notifiability of the message for users.
     all_bot_user_ids = {row["id"] for row in rows if row["is_bot"]}
 
+    # Users who have at least one push device registered to receive push notifications.
+    push_device_registered_user_ids = get_push_device_registered_user_ids(active_user_ids)
```

Comment:
I think it's possible we could avoid the extra DB round-trip by modifying the code just before `rows = list(query)` to have the database do the join internally and just annotate the rows with who has a token.

Reply:
Yeah, I agree. Updated the `UserProfile` query above.